### ISSUE TITLE
Type variable incorrectly bound to entire union when it also appears in a contravariant position

### ISSUE THREAD

**Issue by zhangyoufu** [astral-sh/ty#4446]
### Summary

```python
def fn[T](
    a1: list[T],
    a2: T | list[T],
) -> None: ...

arg: list[int] = [0]
fn(arg, arg)
```

```
error[invalid-argument-type]: Argument to function `fn` is incorrect
 --> reproduce.py:7:4
  |
7 | fn(arg, arg)
  |    ^^^ Expected `list[int | list[int]]`, found `list[int]`
info: element `list[int]` of union `int | list[int]` is not assignable to `int`
info: Function defined here
 --> reproduce.py:1:5
  |
1 | def fn[T](
  |     ^^
2 |     a1: list[T],
  |     ----------- Parameter declared here
info: `list` is invariant in its type parameter
info: Consider using the covariant supertype `collections.abc.Sequence`
info: For more information, see https://docs.astral.sh/ty/reference/typing-faq/#invariant-generics

Found 1 diagnostic
```

### Version

ty 0.0.77

**Comment by sharkdp** [comment]
Thank you for reporting this. I think this is tracked in #2799. I'll add your case as a comment there to make sure it'll get fixed when #2799 is closed.
